api/output_policy.go

```diff
@@ -0,0 +1,197 @@
+	if d.parsedHCLString == "" {
+		d.parseRequest()
+	}
+	return d.parsedHCLString
+}
+
+// Determine whether the given path requires the sudo capability
+func isSudoPath(client *Client, path string) (bool, error) {
+	sudoPaths, err := getSudoPaths(client)
+	if err != nil {
+		return false, fmt.Errorf("unable to retrieve list of paths that require sudo capability: %v", err)
+	}
+	if sudoPaths == nil || len(sudoPaths) < 1 {
+		// OpenAPI spec did not return any paths that require sudo for
+		// some reason, but the user probably still shouldn't see an error.
+		return false, nil
+	}
+
+	// Return early if the path is clearly one of the sudo paths.
+	if _, ok := sudoPaths[path]; ok {
+		return true, nil
+	}
+
+	// Some sudo paths have templated fields in them.
+	// (e.g. sys/revoke-prefix/{prefix})
+	// The keys in the sudoPaths map are actually regular expressions,
+	// so we can check if our path matches against them.
+	for sudoPath := range sudoPaths {
+		r, err := regexp.Compile(fmt.Sprintf("^%s$", sudoPath))
+		if err != nil {
+			continue
+		}
+
+		match := r.Match([]byte(fmt.Sprintf("/%s", path))) // the OpenAPI response has a / in front of each path
+		if match {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func getSudoPaths(client *Client) (map[string]bool, error) {
+	// We don't want to use a wrapping call or any special flags here
+	// so save any custom value and restore after.
+	currentWrappingLookupFunc := client.CurrentWrappingLookupFunc()
+	client.SetWrappingLookupFunc(nil)
+	defer client.SetWrappingLookupFunc(currentWrappingLookupFunc)
+	currentOutputCurlString := client.OutputCurlString()
+	client.SetOutputCurlString(false)
+	defer client.SetOutputCurlString(currentOutputCurlString)
+	currentOutputPolicy := client.OutputPolicy()
+	client.SetOutputPolicy(false)
+	defer client.SetOutputPolicy(currentOutputPolicy)
+
+	r := client.NewRequest("GET", "/v1/sys/internal/specs/openapi")
+	resp, err := client.RawRequest(r)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve sudo endpoints: %v", err)
+	}
+
+	var buf bytes.Buffer
+	_, err = buf.ReadFrom(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body from OpenAPI: %v", err)
+	}
+	if buf.Len() == 0 {
+		return nil, fmt.Errorf("OpenAPI response had no content")
+	}
+
+	oasInfo := make(map[string]interface{})
+	if err := jsonutil.DecodeJSONFromReader(&buf, &oasInfo); err != nil {
+		return nil, fmt.Errorf("unable to decode JSON from OpenAPI response: %v", err)
+	}
+
+	paths, ok := oasInfo["paths"]
+	if !ok {
+		return nil, fmt.Errorf("OpenAPI response did not include paths")
+	}
+
+	pathsMap, ok := paths.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("OpenAPI response did not return valid paths")
+	}
+
+	sudoPaths := make(map[string]bool) // this could be a slice, but we're just making it a map so we can do quick lookups for the paths that don't have any templating
```

Comment:
The performance of looking things up in a slice is actually pretty good due to cache locality. If you have < 15 elements or so, it will likely outperform a map (and has a nicer interface IMO).

Go 1.18 also adds a convenience function [IndexFunc](https://pkg.go.dev/golang.org/x/exp/slices#IndexFunc) for easier slice lookup. Otherwise, we probably have it implemented somewhere in the code.